server/routes/incidentDetails/incidentDetailsPage.ts

```diff
@@ -0,0 +1,728 @@
+  deleteWanted?: string
+  originalPageReferrerUrl?: string
+}
+
+type IncidentDetails = {
+  incidentDate: SubmittedDateTime
+  locationId: number
+  currentIncidentRoleSelection: IncidentRole
+  currentAssociatedPrisonerNumber?: string
+}
+
+type TemporarilySavedData = {
+  originalIncidentRoleSelection: IncidentRole
+  originalReporterUsername: string
+}
+
+type StashedIncidentDetails = {
+  incidentDetails: IncidentDetails
+  temporaryData?: TemporarilySavedData
+}
+
+type ApiIncidentDetails = IncidentDetails & {
+  reporterUsername: string
+}
+
+export enum PageRequestType {
+  CREATION,
+  EDIT,
+  EDIT_SUBMITTED,
+}
+
+enum PageRequestAction {
+  STANDARD,
+  SEARCH_FOR_PRISONER,
+  DELETE_PRISONER,
+}
+
+enum NextPageSelectionAfterEdit {
+  OFFENCE_SELECTION,
+  CHECK_YOUR_ANSWERS,
+  TASK_LIST,
+}
+
+class PageOptions {
+  constructor(private readonly pageType: PageRequestType) {}
+
+  isEdit(): boolean {
+    return this.pageType === PageRequestType.EDIT || this.pageType === PageRequestType.EDIT_SUBMITTED
+  }
+
+  isPreviouslySubmitted(): boolean {
+    return this.pageType === PageRequestType.EDIT_SUBMITTED
+  }
+}
+
+export default class IncidentDetailsPage {
+  pageOptions: PageOptions
+
+  constructor(
+    pageType: PageRequestType,
+    private readonly placeOnReportService: PlaceOnReportService,
+    private readonly locationService: LocationService
+  ) {
+    this.pageOptions = new PageOptions(pageType)
+  }
+
+  view = async (req: Request, res: Response): Promise<void> => {
+    const requestValues = extractValuesFromRequest(req)
+    const requestData = getPageActionFromRequest(requestValues)
+    const { user } = res.locals
+    debugData('GET', requestValues)
+    debugData('GET action', requestData)
+
+    let data: StashedIncidentDetails = null
+    if (requestData !== PageRequestAction.STANDARD) {
+      const sessionData = popDataFromSession(req)
+      if (requestData === PageRequestAction.SEARCH_FOR_PRISONER) {
+        data = updateDataOnSearchReturn(sessionData, requestValues)
+      } else if (requestData === PageRequestAction.DELETE_PRISONER) {
+        data = updateDataOnDeleteReturn(sessionData, requestValues)
+      }
+    } else if (this.pageOptions.isEdit()) {
+      const draftIncidentDetails = await this.readFromApi(requestValues.draftId, user as User)
+      data = {
+        incidentDetails: draftIncidentDetails,
+        temporaryData: {
+          originalIncidentRoleSelection: draftIncidentDetails.currentIncidentRoleSelection,
+          originalReporterUsername: draftIncidentDetails.reporterUsername,
+        },
+      }
+    }
+    debugData('GET DATA', data)
+
+    let originalReporterUsernameForPage = user.username
+    if (this.pageOptions.isEdit()) {
+      originalReporterUsernameForPage = data.temporaryData?.originalReporterUsername
+    }
+    const pageData = await this.getPageDataOnGet(requestValues, originalReporterUsernameForPage, data, user)
+    renderData(res, pageData, null)
```

Comment:
Normally we return renderData as a call? Maybe it doesn't matter? Just wondering if the deleteSessionData should be called before this and then we return it?